types/ramda/index.d.ts

```diff
@@ -173,7 +173,11 @@ export function allPass<T, TF1 extends T, TF2 extends T, TF3 extends T, TF4 exte
         PredTypeguard<T, TF6>,
     ],
 ): PredTypeguard<T, TF1 & TF2 & TF3 & TF4 & TF5 & TF6>;
-export function allPass<F extends Pred>(preds: readonly F[]): F;
+// For inference, so should be first. Works with generics and overloads
+export function allPass<Pred extends (...args: any[]) => unknown, _Overload extends 'pred' = 'pred'>(preds: readonly Pred[]): Pred;
```

Comment:
What is `_Overload` doing, since it isn't used?

Reply:
Check the tests - it's added so that you can select the overload. In this case it's not useful at all, it's just on this specific overload for consistency

Really, the only real usecase is if you want to pass a tuple as a single arg - it would normally be caught by the `args` overload, but if you manually specify `'arg'` you'd get the desired behavior.

I do think it's not very useful for `allPass` and `anyPass` specifically so it's still fine to reject this PR
I think there are some other places where it *would* be potentially useful - seems like in the overhaul PR I added it so you would be able to select which of the array/record/functor overloads to use, since they all take the same type arguments (input and output)

Reply:
My main concern with this is, that adding Generics like that to "select" an overload is not really documented TypeScript behaviour. So even if it would be useful, very few people would use it, because no one would know its there. Usually I don't want to look at the actual type of a function to work with it, which would be necessary to discover this "feature".

I also think its potentially confusing for maintainers as well.

Or did I misunderstand something here?

Reply:
That's a very good point, and yeah... I don't know. I did mention in the initial message that I'm mostly just trying to figure out whether this is a good idea in the first place.

Indeed it isn't documented TypeScript behaviour, but worth noting that ts-toolbelt uses something similar. Not that ts-toolbelt should be considered a good reference...

I guess what it boils down to is, is there any situation in which adding this would be useful?
I'd say... *maybe*, I don't really know - there are a ton of overloads, some of which have the same argument lists

I guess I forgot to mention, here's the implementation that was in the overhaul PR:
```typescript
export function map<_Overload extends 'array', T, U>(fn: (x: T) => U): (list: readonly T[]) => U[];
export function map<_Overload extends 'record', T, U>(fn: (x: T) => U): <K extends string>(list: Record<K, T>) => Record<K, U>;
export function map<_Overload extends 'functor', T, U>(fn: (x: T) => U): (functor: Functor<T>) => Functor<U>;
```
While I'm not 100% sure, I suspect this would come in useful when passing `map` into any of the higher order functions, otherwise IIRC `map` can only work for one type - currently that is arrays, since they're the bottom-most overload in the list.
In other words, it's currently impossible to specify "I want to use `.map(fn): (functor: Functor<T>) => Functor<U>`" when you're passing it to say, `pipe`, or `converge`

Reply:
I think TypeScript is totally capable of typing this. Didn't test it, but something like this should be fine, and type inference in `pipe` should work (if `pipe` is typed correctly, which I actually didn't check):

```typescript
export function map<T, U>(fn: (x: T) => U): {
    (list: readonly T[]): U[],
    <K extends string>(list: Record<K, T>): Record<K, U>,
    (functor: Functor<T>): Functor<U>
}
```
[Playground](https://www.typescriptlang.org/play?ssl=5&ssc=2&pln=1&pc=1#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwLYCGYAPACoA0cAqgHwAUCKAXHHSC6QJRwC8N1nFgG8AUHHGsANpgDOMFlGAEAJjkkBPOKQDaAXUHU95MROIBpOKBjAUymXDlRMKAOb1pclgCVgaaMvNKUhoDHz8oALNKWmMJVmR0GGgWADFUDGgyENT0pKhiWhEAXxERPxQ5fCRJGAAmXnwiOjoCFhQkPAAjYChuPjgCADokgGUYJ1c6Tk4AblLyyus5AEEoKAJNHiqa2rptAEZyWvIAZn0ynEXgOTD-BrxqurohVrhDuE6WY7g0FhOizhAA)

Reply:
Also I really like the structure this gives, because it groups overloads, that start out the same.

Reply:
I do believe `pipe` does, in fact, not work:
[Playground](https://www.typescriptlang.org/play?ssl=15&ssc=9&pln=14&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAIimRABNlCBuAKBoFMAPSWHAVwDsBjGYCDuCGRgAPABUANHACqAPgAU2DgC448xqrEBKOAF5ZMraoDeNOObUAbYKhiqo9ZGX6WAnnDEBtALpGZPiTMLEQBpOCYYeg4ydFsoYA4AcwVrW1UMei5oMlCpMVk-DKyoHJCpOUCLNWxOHmhVADFamGhxAsbm1rkaAF86LI5bQTZLGAAmPUFheXlkVQ42EAAjeigdfRQAOhaAZRh4pPktLVoaAaHI2wBBKBJ3XWHRsflPAEYJMYkAZl8z-gv6LYitlJiARuN5MY5nB3nAlqpPnAuKovj0tP1-vAIAALDgQSYYTZgYBgejyIRgGbQhbLVbrAzIbYQPYHRJHY60AD0nKqAD0APxAA)

In this PR I've changed the type of `pipe` to work nicely for one parameter, but it still will not work for more than one parameter.

Reply:
At work we actually use our own version of `pipe` which works great. I'll take a look at Ramda `pipe` the other day, because its definitely typeable.

Reply:
re: the overload grouping - that still exists. In the case of `map`, those are extra overloads.
Agree that the overload returning an overloaded function should exist and is a huge boon to usability when `map` is used at the top level